File: malipo-flutter/lib/src/resources/webhooks.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import '../models.dart';

/// Default replay-protection window, mirroring the Node SDK (5 minutes).
const int _defaultToleranceMs = 300000;

class WebhooksResource {
  /// Verify a webhook signature and return the decoded event.
  ///
  /// The signed message is `<timestamp>.<payload>`: the timestamp is prefixed to the
  /// exact raw body with a literal dot, matching the Node SDK. When [timestamp] is
  /// omitted the legacy body-only scheme is used.
  ///
  /// [payload] is the raw request body, exactly as received. Do not re-serialize it.
  /// [signature] is the value of the `X-Webhook-Signature` header.
  /// [secret] is your webhook signing secret (64 hex characters, no prefix).
  /// [timestamp] is the value of the `X-Webhook-Timestamp` header. Required for
  /// deliveries signed with the timestamp scheme.
  /// [toleranceMs] is the replay-protection window in milliseconds (default 300000,
  /// i.e. 5 minutes). Set to 0 to disable the check, for example when verifying a
  /// frozen test vector.
  MalipoEvent constructEvent(
    String payload,
    String signature,
    String secret, [
    String? timestamp,
    int toleranceMs = _defaultToleranceMs,
  ]) {
    if (payload.isEmpty || signature.isEmpty || secret.isEmpty) {
      throw ArgumentError('Missing payload, signature, or secret for webhook verification.');
    }

    final hasTimestamp = timestamp != null && timestamp.isNotEmpty;

    if (hasTimestamp && toleranceMs > 0) {
      final signedAt = DateTime.tryParse(timestamp);

      if (signedAt == null ||
          DateTime.now().toUtc().difference(signedAt.toUtc()).inMilliseconds.abs() > toleranceMs) {
        throw Exception('Webhook timestamp out of window.');
      }
    }

    final signedPayload = hasTimestamp ? '$timestamp.$payload' : payload;
    final expectedSignature =
        Hmac(sha256, utf8.encode(secret)).convert(utf8.encode(signedPayload)).toString();

    if (!_constantTimeEquals(expectedSignature, signature)) {
      throw Exception('Invalid webhook signature.');
    }

    try {
      final decodedPayload = jsonDecode(payload) as Map<String, dynamic>;
      return MalipoEvent.fromJson(decodedPayload);
    } catch (e) {
      throw Exception('Failed to decode webhook payload: $e');
    }
  }

  /// Length-safe comparison that does not short-circuit on the first differing byte.
  static bool _constantTimeEquals(String expected, String provided) {
    final expectedBytes = utf8.encode(expected);
    final providedBytes = utf8.encode(provided);

    if (expectedBytes.length != providedBytes.length) {
      return false;
    }

    var difference = 0;
    for (var i = 0; i < expectedBytes.length; i++) {
      difference |= expectedBytes[i] ^ providedBytes[i];
    }

    return difference == 0;
  }
}

File: malipo-php/src/Webhooks.php
<?php

namespace Malipo;

use Malipo\Exceptions\MalipoException;

class Webhooks
{
    /** Default replay-protection window, mirroring the Node SDK (5 minutes). */
    public const DEFAULT_TOLERANCE_MS = 300000;

    /**
     * Verify a webhook signature and return the decoded event.
     *
     * The signed message is the timestamp, a literal dot, then the exact raw body,
     * matching the Node SDK. When no timestamp is supplied the legacy body-only
     * scheme is used.
     *
     * @param string      $payload     The raw request body, exactly as received. Do not re-serialize it.
     * @param string      $signature   The value of the X-Webhook-Signature header.
     * @param string      $secret      Your webhook signing secret (64 hex characters, no prefix).
     * @param string|null $timestamp   The value of the X-Webhook-Timestamp header.
     * @param int         $toleranceMs Replay-protection window in milliseconds. Set to 0 to disable.
     *
     * @return array<string, mixed>
     */
    public function constructEvent(
        string $payload,
        string $signature,
        string $secret,
        ?string $timestamp = null,
        int $toleranceMs = self::DEFAULT_TOLERANCE_MS
    ) {
        if (empty($payload) || empty($signature) || empty($secret)) {
            throw new \InvalidArgumentException("Missing payload, signature, or secret for webhook verification.");
        }

        $hasTimestamp = $timestamp !== null && $timestamp !== '';

        if ($hasTimestamp && $toleranceMs > 0) {
            $signedAt = strtotime($timestamp);

            if ($signedAt === false || abs((microtime(true) * 1000) - ($signedAt * 1000)) > $toleranceMs) {
                throw new MalipoException("Webhook timestamp out of window.", 400);
            }
        }

        $signedPayload = $hasTimestamp ? $timestamp . '.' . $payload : $payload;
        $expectedSignature = hash_hmac('sha256', $signedPayload, $secret);

        if (!hash_equals($expectedSignature, $signature)) {
            throw new MalipoException("Invalid webhook signature.", 400);
        }

        return json_decode($payload, true);
    }
}
